x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```diff
@@ -172,4 +210,50 @@ private void removeSessionForShard(String sessionUUID, IndexShard indexShard) {
             }
         }
     }
+
+    private static class RefCountedCloseable<T extends Closeable> extends AbstractRefCounted {
+
+        private static final String NAME = "ref-counted-session-object";
+
+        private final String name;
+        private final T object;
+
+        private RefCountedCloseable(String name, T object) {
+            super(NAME);
+            this.name = name;
+            this.object = object;
+        }
+
+        @Override
+        protected void closeInternal() {
+            IOUtils.closeWhileHandlingException(object);
+        }
+    }
+
+    public static class Reader implements Closeable {
+
+        private final RefCountedCloseable<Engine.IndexCommitRef> session;
+        private final RefCountedCloseable<IndexInput> input;
+
+        private Reader(RefCountedCloseable<Engine.IndexCommitRef> session, RefCountedCloseable<IndexInput> input) {
+            this.session = session;
+            this.input = input;
+            input.incRef();
+            session.incRef();
+        }
+
+        public void readFileBytes(byte[] chunk, long offset, int length) throws IOException {
+            synchronized (input.object) {
```

Comment:
see javadocs of IndexInput:
> <p>{@code IndexInput} may only be used from one thread, because it is not
> * thread safe (it keeps internal state like file position). To allow
> * multithreaded use, every {@code IndexInput} instance must be cloned before
> * it is used in another thread. Subclasses must therefore implement {@link #clone()},
> * returning a new {@code IndexInput} which operates on the same underlying
> * resource, but positioned independently.

Reply:
I made this change:

```
            synchronized (input.object) {
                try (IndexInput in = input.object.clone()) {
                    in.seek(offset);
                    in.readBytes(chunk, 0, length);
                }
            }
```

I think that is correct?